src/main/java/org/folio/service/impl/PatronPinService.java

```diff
@@ -0,0 +1,34 @@
+package org.folio.service.impl;
+
+import static javax.xml.bind.DatatypeConverter.printHexBinary;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+
+import javax.crypto.SecretKey;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PatronPinService {
+  private static final Logger logger = LogManager.getLogger(PatronPinService.class);
+
+  private final PasswordHashService passwordHashService;
+
+  public PatronPinService(PasswordHashService passwordHashService) {
+    this.passwordHashService = passwordHashService;
+  }
+
+  public String derivePin(String pin, String userId) {
+    try {
+      // Should the salt also include the tenant?
```

Comment:
Salts should always, in theory, be random (but saved somewhere) so that they act to further obscure a hash.
By adding a known, such as a username or in this case a tenant, it is rather self-defeating.
The theoretical attacker would easly be able to know the "tenant" part, thereby reducing the strength of the salt.

For example a salt of `12bob`, with `bob` being the tenant, would effectively be the same has having a salt of `12`.

With that said, having two knowns (userId and Tenant) is probably more secure than one known (userId).
At least the ordering can offer some slight security.

I would recommend (somehow) having the salt generated and saved, such as a configuration file that loads the salt and then at some point this random salt is passed to this function. I believe there are some implementations that store this data in the database rather than on the disk. The trick is figuring out the safest way to store the salt while still being able to fetch and use it in a way that works with a given project. More secure salts would have a different salt per user, but this approach would require more storage and fetching.

Reply:
TLDR: I purposely didn't change the decision that had already been made to use the user ID as the salt.

I think feedback about whether this is an appropriate salt would be better given to @ianibo and the Leipzig folks who asked for this functionality, the original issues for this work can be found [here|https://issues.folio.org/browse/MODUSERS-254].

This code could be switched to use a secure random number generator and store the salt with the pin.

It would mean needing to migrate the existing data during an upgrade. And due to limitations with how the data model and API record representation are the same in RMB modules, would need some work to make sure it wasn't exposed to a client.

> Salts should always, in theory, be random (but saved somewhere) so that they act to further obscure a hash. By adding a known, such as a username or in this case a tenant, it is rather self-defeating. The theoretical attacker would easly be able to know the "tenant" part, thereby reducing the strength of the salt.

I agree that a cryptographically random number could be a better choice of salt than a known value.

My cryptography is pretty rusty, based upon this [article|https://auth0.com/blog/adding-salt-to-hashing-a-better-way-to-store-passwords/] I think the two goals of salting (getting a different hash for the same pin and not being able to use a shared table for table attacks) are still achieved by using a different hash for each user, even if it isn't the optimal solution (because it could allow for different password / pin fields with the same value, to end up with the same hash, which could reveal information to an attacker.

> For example a salt of `12bob`, with `bob` being the tenant, would effectively be the same has having a salt of `12`.
> 
> With that said, having two knowns (userId and Tenant) is probably more secure than one known (userId). At least the ordering can offer some slight security.
> 
> I would recommend (somehow) having the salt generated and saved, such as a configuration file that loads the salt and then at some point this random salt is passed to this function. I believe there are some implementations that store this data in the database rather than on the disk. The trick is figuring out the safest way to store the salt while still being able to fetch and use it in a way that works with a given project. More secure salts would have a different salt per user, but this approach would require more storage and fetching.

As I understand it, salts are also typically intended to be different for each key, as this contributes to